File: backend/app/schemas/screener.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class HardFilterPrefs(BaseModel):
    exclude_st: bool = True
    exclude_suspended: bool = True
    min_amount_wan: float = 3000.0
    min_total_mv_yi: float = 50.0
    exclude_new_listing: bool = False
    min_listing_days: int = 60
    exclude_limit_board: bool = False
    exclude_one_word: bool = False
    allowed_industries: str = ""
    allowed_market_boards: str = ""


class HardFilterTemplate(BaseModel):
    id: str
    name: str
    prefs: HardFilterPrefs


class IndustryListOut(BaseModel):
    items: list[str]


class PresetOut(BaseModel):
    name: str
    source: str
    rule_kind: str
    description: str
    implemented: bool = True


class SchemeCreate(BaseModel):
    name: str
    config: dict[str, Any] = Field(default_factory=dict)


class SchemeUpdate(BaseModel):
    name: str | None = None
    config: dict[str, Any] | None = None


class SchemeOut(BaseModel):
    id: str
    name: str
    config: dict[str, Any]
    created_at: str
    updated_at: str


class RecipeCreate(BaseModel):
    name: str
    trigger_kind: str = "intraday"
    config: dict[str, Any] = Field(default_factory=dict)


class RecipeUpdate(BaseModel):
    name: str | None = None
    trigger_kind: str | None = None
    config: dict[str, Any] | None = None


class RecipeOut(BaseModel):
    id: str
    name: str
    trigger_kind: str
    config: dict[str, Any]
    created_at: str
    updated_at: str


class BuiltinRecipeOut(BaseModel):
    recipe_id: str
    name: str
    trigger_kind: str
    top_n: int
    implemented: bool


class PatternOut(BaseModel):
    pattern_id: str
    name: str
    description: str


class PatternRunRequest(BaseModel):
    pattern_id: str = Field(description="ma_bull | w_bottom | old_duck | theme_hot")
    top_n: int = Field(default=20, ge=1, le=100)
    max_scan: int = Field(default=800, ge=50, le=1200)
    hard_filter: HardFilterPrefs | None = None
    hard_filter_template: str | None = None


class ReferencePeerRequest(BaseModel):
    vt_symbol: str = Field(min_length=1, description="标杆股，如 600519.SSE")
    top_n: int = Field(default=20, ge=1, le=100)
    reference_name: str = ""
    hard_filter: HardFilterPrefs | None = None
    hard_filter_template: str | None = None
    weights: dict[str, float] | None = None


class ConditionRunRequest(BaseModel):
    preset: str = "涨幅榜"
    top_n: int = Field(default=50, ge=1, le=500)
    hard_filter: HardFilterPrefs | None = None
    hard_filter_template: str | None = None
    min_change_pct: float | None = None
    max_change_pct: float | None = None
    min_turnover_rate: float | None = None
    max_turnover_rate: float | None = None


class RecipeRunRequest(BaseModel):
    recipe_id: str = "intraday_multi"
    top_n: int | None = Field(default=None, ge=1, le=200)
    hard_filter: HardFilterPrefs | None = None
    hard_filter_template: str | None = None
    variant: str | None = Field(default=None, description="radar_leader: mainline | all_market")


class RecipeWeightItem(BaseModel):
    key: str
    label: str
    weight: float
    default_weight: float


class RecipeWeightsOut(BaseModel):
    recipe_id: str
    items: list[RecipeWeightItem]
    weights: dict[str, float]


class RecipeWeightsPut(BaseModel):
    weights: dict[str, float] = Field(default_factory=dict)


class JobAccepted(BaseModel):
    job_id: str


class RunSummary(BaseModel):
    id: str
    condition: str
    source: str
    row_count: int
    total_scanned: int
    created_at: str


class RunDetail(RunSummary):
    config: dict[str, Any]
    result: dict[str, Any]


class JobOut(BaseModel):
    id: str
    kind: str
    status: str
    progress: float = 0.0
    error: str | None = None
    result_ref: str | None = None
    created_at: str
    updated_at: str

File: frontend/src/api/screener.ts
import { api, pageQuery, type Page } from './client'

export type User = {
  id: string
  username: string
  display_name: string
}

export type TokenResponse = {
  access_token: string
  token_type: string
  user: User
}

export type Preset = {
  name: string
  source: string
  rule_kind: string
  description: string
  implemented: boolean
}

export type HardFilterPrefs = {
  exclude_st: boolean
  exclude_suspended: boolean
  min_amount_wan: number
  min_total_mv_yi: number
  exclude_new_listing: boolean
  min_listing_days: number
  exclude_limit_board: boolean
  exclude_one_word: boolean
  allowed_industries: string
  allowed_market_boards: string
}

export type HardFilterTemplate = {
  id: string
  name: string
  prefs: HardFilterPrefs
}

export type BuiltinRecipe = {
  recipe_id: string
  name: string
  trigger_kind: string
  top_n: number
  implemented: boolean
}

export type PatternMeta = {
  pattern_id: string
  name: string
  description: string
}

export type Scheme = {
  id: string
  name: string
  config: Record<string, unknown>
  created_at: string
  updated_at: string
}

export type Job = {
  id: string
  kind: string
  status: string
  progress: number
  error: string | null
  result_ref: string | null
  created_at: string
  updated_at: string
}

export type RunSummary = {
  id: string
  condition: string
  source: string
  row_count: number
  total_scanned: number
  created_at: string
}

export type RunDetail = RunSummary & {
  config: Record<string, unknown>
  result: {
    rows?: Record<string, unknown>[]
    industry_dist?: { industry: string; count: number; ratio: number }[]
    diff?: { added: string[]; removed: string[]; kept: string[] }
    [key: string]: unknown
  }
}

export type RecipeWeightItem = {
  key: string
  label: string
  weight: number
  default_weight: number
}

export type RecipeWeights = {
  recipe_id: string
  items: RecipeWeightItem[]
  weights: Record<string, number>
}

export const authApi = {
  login: (username: string, password: string) =>
    api<TokenResponse>('/api/v1/auth/login', {
      method: 'POST',
      body: JSON.stringify({ username, password }),
    }),
  me: () => api<User>('/api/v1/auth/me'),
}

export const screenerApi = {
  presets: () => api<Preset[]>('/api/v1/screener/presets'),
  industries: () => api<{ items: string[] }>('/api/v1/screener/industries'),
  hardFilterTemplates: () => api<HardFilterTemplate[]>('/api/v1/screener/hard-filter-templates'),
  builtinRecipes: () => api<BuiltinRecipe[]>('/api/v1/screener/builtin-recipes'),
  patterns: () => api<PatternMeta[]>('/api/v1/screener/patterns'),
  dataStatus: () =>
    api<{
      redis: { available: boolean; quote_count: number; updated_at: string | null }
      tushare_configured: boolean
    }>('/api/v1/screener/data-status'),
  schemes: () => api<Scheme[]>('/api/v1/screener/schemes'),
  createScheme: (name: string, config: Record<string, unknown>) =>
    api<Scheme>('/api/v1/screener/schemes', {
      method: 'POST',
      body: JSON.stringify({ name, config }),
    }),
  deleteScheme: (id: string) =>
    api<{ ok: boolean }>(`/api/v1/screener/schemes/${encodeURIComponent(id)}`, {
      method: 'DELETE',
    }),
  runCondition: (body: Record<string, unknown>) =>
    api<{ job_id: string }>('/api/v1/screener/runs/condition', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  runRecipe: (body: Record<string, unknown>) =>
    api<{ job_id: string }>('/api/v1/screener/runs/recipe', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  runPattern: (body: Record<string, unknown>) =>
    api<{ job_id: string }>('/api/v1/screener/runs/pattern', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  runReferencePeer: (body: Record<string, unknown>) =>
    api<{ job_id: string }>('/api/v1/screener/runs/reference-peer', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  runsPage: (page = 1, pageSize = 20) =>
    api<Page<RunSummary>>(`/api/v1/screener/runs/page?${pageQuery(page, pageSize)}`),
  run: (id: string) => api<RunDetail>(`/api/v1/screener/runs/${id}`),
  exportCsvUrl: (id: string) => `/api/v1/screener/runs/${id}/export.csv`,
  recipeWeights: (recipeId: string) =>
    api<RecipeWeights>(`/api/v1/screener/recipes/${encodeURIComponent(recipeId)}/weights`),
  putRecipeWeights: (recipeId: string, weights: Record<string, number>) =>
    api<RecipeWeights>(`/api/v1/screener/recipes/${encodeURIComponent(recipeId)}/weights`, {
      method: 'PUT',
      body: JSON.stringify({ weights }),
    }),
}

export const jobsApi = {
  get: (id: string) => api<Job>(`/api/v1/jobs/${id}`),
}
